src/libtiled/mapreader.cpp

```diff
@@ -825,42 +821,96 @@ void MapReaderPrivate::readTileLayerData(TileLayer &tileLayer,
 
     mMap->setLayerDataFormat(layerDataFormat);
 
-    int x = 0;
-    int y = 0;
-
-    while (xml.readNext() != QXmlStreamReader::Invalid) {
-        if (xml.isEndElement()) {
-            break;
-        } else if (xml.isStartElement()) {
-            if (xml.name() == QLatin1String("tile")) {
-                if (y >= tileLayer.height()) {
-                    xml.raiseError(tr("Too many <tile> elements"));
-                    continue;
+    if (mMap->infinite()) {
+        while (xml.readNext() != QXmlStreamReader::Invalid) {
+            if (xml.isEndElement()) {
+                break;
+            } else if (xml.isStartElement()) {
+                if (xml.name() == QLatin1String("chunk")) {
+                    const QXmlStreamAttributes atts = xml.attributes();
+                    int startX = atts.value(QLatin1String("startx")).toInt();
+                    int startY = atts.value(QLatin1String("starty")).toInt();
+
+                    int x = 0;
+                    int y = 0;
+
+                    while (xml.readNext() != QXmlStreamReader::Invalid) {
+                        if (xml.isEndElement()) {
+                            break;
+                        } else if (xml.isStartElement()) {
+                            if (xml.name() == QLatin1String("tile")) {
+                                if (y >= CHUNK_SIZE) {
+                                    xml.raiseError(tr("Too many <tile> elements"));
+                                    continue;
+                                }
+
+                                const QXmlStreamAttributes atts = xml.attributes();
+                                unsigned gid = atts.value(QLatin1String("gid")).toUInt();
+                                tileLayer.setCell(x + startX, y + startY, cellForGid(gid));
+
+                                x++;
+                                if (x >= CHUNK_SIZE) {
+                                    x = 0;
+                                    y++;
+                                }
+
+                                xml.skipCurrentElement();
+                            } else {
+                                readUnknownElement();
+                            }
+                        } else if (xml.isCharacters() && !xml.isWhitespace()) {
+                            if (encoding == QLatin1String("base64")) {
+                                decodeBinaryLayerData(tileLayer,
+                                                      xml.text().toLatin1(),
+                                                      layerDataFormat,
+                                                      startX,
+                                                      startY);
+                            } else if (encoding == QLatin1String("csv")) {
+                                decodeCSVLayerData(tileLayer, xml.text(), startX, startY);
+                            }
+                        }
+                    }
                 }
+            }
+        }
```

Comment:
I think it would make sense to move this (and the following else branch) into a seperate function to make the code a bit easier to read.

That many braces closing in the same place is usually a good indicator that the code could be moved into a function...

Reply:
Yes, the reading of the inner part of either `data` itself or each of its `chunk` elements should be done in a separate function rather than duplicating the whole code. Property this function again needs to take a target `QRect` so that it knows where to set the tiles.